controller/src/main/java/io/pravega/controller/store/PravegaTablesStoreHelper.java

```diff
@@ -244,6 +248,101 @@ public void invalidateCache(String table, String key) {
+                                                        // Delete TagRecords with empty streams.
+                                                        .thenCompose(keyVersions -> deleteEmptyTagRecords(tableName, requestId, result.getRight(), result.getLeft(), keyVersions))
+                                                        .whenComplete((v, ex) -> {
+                                                            releaseEntries(result.getLeft());
+                                                        });
+                                });
+        }, errorMessage, true, requestId);
+    }
+
+    private CompletableFuture<Void> deleteEmptyTagRecords(String tableName, long requestId, List<Integer> deleteIndex,
+                                                          List<TableSegmentEntry> updatedEntryList,
+                                                          List<TableSegmentKeyVersion> keyVersions) {
+        if (deleteIndex.isEmpty()) {
+            return CompletableFuture.completedFuture(null);
+        }
+        // Compute the list of keys to be deleted.
+        List<TableSegmentKey> conditionalDeleteKeys = deleteIndex.stream()
+                                                                 .map(index -> {
+                                                                     ByteBuf keyBuf = updatedEntryList.get(index).getKey().getKey();
+                                                                     TableSegmentKeyVersion version = keyVersions.get(index);
+                                                                     return TableSegmentKey.versioned(keyBuf, version.getSegmentVersion());
+                                                                 }).collect(Collectors.toList());
+        Supplier<String> errorMessage = () -> String.format("cleanTagEntries: table: %s", tableName);
+        // attempt to do a conditional delete of the keys.
+        return withRetries(() -> segmentHelper.removeTableKeys(tableName, conditionalDeleteKeys, authToken.get(), requestId), errorMessage, requestId)
+                .exceptionally(t -> {
+                    log.info(requestId, "Clean Tag Entries failed, ignoring this exception {}", t.getMessage());
+                    return null;
+                }).whenComplete((v, t) -> releaseKeys(conditionalDeleteKeys));
+    }
+
+    /**
+     * Append value to every entry corresponding to the provided keys.
+     *
+     * @param tableName      Table name.
+     * @param tableKeys      List of keys whose value needs to be updated.
+     * @param appendValue    Value to be appended to all the keys.
+     * @param updateFunction Update function which is used to compute the updated entries.
+     * @param requestId      Request id.
+     * @return Future which completes when the operation is complete.
+     */
+    public CompletableFuture<List<Version>> appendValues(final String tableName, Set<String> tableKeys, final String appendValue,
```

Comment:
same here. 
add an updateValues method in helper. which your code can then make use of.

Reply:
Done, Improved the abstraction.